File: cli/template/packages/java/src/main/java/do_/{{name}}/{{Name}}Client.java
package do_.{{name}};

import do_.rpc.RpcClient;
import do_.rpc.RpcConnection;

/**
 * {{Name}}.do SDK client.
 *
 * <p>Example usage:
 * <pre>{@code
 * var client = new {{Name}}Client(System.getenv("DOTDO_KEY"));
 * var rpc = client.connect();
 * var result = rpc.call("example");
 * }</pre>
 */
public class {{Name}}Client implements AutoCloseable {
    private static final String DEFAULT_BASE_URL = "https://{{name}}.do";

    private final String apiKey;
    private final String baseUrl;
    private RpcClient rpc;

    /**
     * Create a new client with an API key.
     *
     * @param apiKey the API key for authentication
     */
    public {{Name}}Client(String apiKey) {
        this(apiKey, DEFAULT_BASE_URL);
    }

    /**
     * Create a new client with an API key and custom base URL.
     *
     * @param apiKey the API key for authentication
     * @param baseUrl the base URL for the service
     */
    public {{Name}}Client(String apiKey, String baseUrl) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl;
    }

    /**
     * Connect to the {{name}}.do service.
     *
     * @return the connected RPC client
     * @throws {{Name}}Exception if connection fails
     */
    public RpcClient connect() throws {{Name}}Exception {
        if (rpc != null) {
            return rpc;
        }

        try {
            var connection = RpcConnection.builder()
                .url(baseUrl)
                .header("Authorization", "Bearer " + apiKey)
                .build();
            rpc = new RpcClient(connection);
            return rpc;
        } catch (Exception e) {
            throw new {{Name}}Exception("Failed to connect", e);
        }
    }

    /**
     * Get the base URL.
     *
     * @return the base URL
     */
    public String getBaseUrl() {
        return baseUrl;
    }

    /**
     * Get the API key.
     *
     * @return the API key
     */
    public String getApiKey() {
        return apiKey;
    }

    @Override
    public void close() {
        if (rpc != null) {
            rpc.close();
            rpc = null;
        }
    }
}

File: cli/template/packages/scala/src/main/scala/do_/{{name}}/{{Name}}Client.scala
package do_.{{name}}

import cats.effect.IO
import do_.rpc.RpcClient
import do_.rpc.RpcConnection

/**
 * {{Name}}.do SDK Client
 *
 * {{description}}
 *
 * @example
 * {{{
 * val client = {{Name}}Client(apiKey = sys.env.get("DOTDO_KEY"))
 * val result = client.connect().flatMap(_.call("example"))
 * }}}
 */
case class {{Name}}ClientOptions(
  apiKey: Option[String] = None,
  baseUrl: String = "https://{{name}}.do"
)

class {{Name}}Client(options: {{Name}}ClientOptions = {{Name}}ClientOptions()) {
  private var rpc: Option[RpcClient] = None

  /**
   * Connect to the {{name}}.do service
   */
  def connect(): IO[RpcClient] = {
    rpc match {
      case Some(client) => IO.pure(client)
      case None =>
        val headers = options.apiKey.map(key =>
          Map("Authorization" -> s"Bearer $key")
        ).getOrElse(Map.empty)

        RpcConnection.connect(options.baseUrl, headers).map { client =>
          rpc = Some(client)
          client
        }
    }
  }

  /**
   * Disconnect from the service
   */
  def disconnect(): IO[Unit] = IO {
    rpc = None
  }

  /**
   * Get the base URL
   */
  def baseUrl: String = options.baseUrl
}

object {{Name}}Client {
  def apply(apiKey: Option[String] = None, baseUrl: String = "https://{{name}}.do"): {{Name}}Client =
    new {{Name}}Client({{Name}}ClientOptions(apiKey, baseUrl))
}
